Add support for incoming webhooks

ttuffin opened the pull request:
Closes #1.

First iteration of webhook server code and webhook plugin example. Includes updates to settings.py to include config for webhook server. I also updated docker-compose.yml to enable host network mode to allow communication between docker image and bot running on localhost.

For the integration to work you need to allow untrusted internal connections to the localhost, e.g `127.0.0.1 0.0.0.0 localhost ::1`. This is done via System Console -> Environment -> Developer

@jneeven let me know if this is heading in the right direction. I would like your thoughts on how/where we process webhooks. At the moment there is just basic processing in process_webhook() and process_action(). Do we want to just keep all the processing here, or would it be better to have these functions call methods in another module? 

Todo:

- [x] Run webhook server on a dedicated thread (@ttuffin)
- [x] Add a queue and a queue "listener" to `MessageHandler` and have the webhook server adds things to the queue (@ttuffin)
- [x] Add `listen_webhook` decorator, rename `Function` to `MessageFunction` and create new subclass `WebhookFunction`  (@ttuffin / @jneeven)
- [x] Add `MessageHandler._handle_webhook` function similar to `_handle_message`, calling `Plugin.call_function`. (@jneeven)
- [x] Add some typical webhook-related functions to the `Driver` class (@ttuffin)
- [x] Create `WebHookEvent` object similar to `Message` and have `Plugin.call_function` support that as well (@jneeven)
- [ ] Implement logging.
- [x] Add shutdown and cleanup routine.
- [x] Write tests.

jneeven commented:
@ttuffin this looks great, thanks for getting this off the ground! I've tested it locally and this could really take the functionality to the next level. I'll modify the docker image to include that connection setting, then we can eventually create integration tests for this as well.

My idea is as follows:
- The webhook server simply listens to any hook_id, exactly as you've currently implemented it, on a separate thread or process (I believe it currently runs in the main thread?)
- We'll add a function `_handle_webhook` to the `MessageHandler`, much like there currently is `_handle_post`, except that it will get its events from a queue rather than from the mattermost socket connection.
- Whenever the webhook server gets a hook request, it simply adds the id and payload to the queue, so that the `MessageHandler` will take care of the rest.
- We add another decorator, e.g. `listen_webhook(<your_id_here>)`, which can be used to decorate the class methods of Plugins
- The `MessageHandler._handle_webhook` function will retrieve the webhook event from the queue, match it to the listeners defined on the active plugins, and trigger `call_function` (which will need to be slightly modified to distinguish message triggers and webhook triggers) similarly to how it currently handles messages.
- The function registered on the plugin will be very similar to functions we currently have, except it'll take a either a JSON dict or a `WebhookEvent` object of some kind instead of a `Message` object.
- The response code you currently have in `process_webhook`:
```python
    async with ClientSession() as session:
        await session.post(webhook_url, json=payload)
 ```
 Can become a function on the driver class, similar to how there currently is `Driver.reply_to(Message)` but then it'll take a webhook event instead.
 
This still leaves a bunch of implementation details to be filled in later, but this is the rough idea I had. With the description above I don't think we'd need to distinguish between actions and hooks, but I'm not sure. In the worst case we can implement an `Action` object as well and create some dedicated functions for that too.

Eventually it would be nice if we would generate webhook IDs and register them on the fly, such that there could be a stateful interaction of some kind (e.g. the first button would trigger a certain webhook, which would send another form as a response, in turn triggering another freshly generated webhook registered to the same function, so that the function knows at what stage of the interaction it's in), but we can worry about that in a much later stage (probably after integrating this with mmpy_bot).

Let me know what you think!

ttuffin commented:
Thanks @jneeven for the detailed information! I'll likely have more questions later on, but this is great info for me to continue with.

I agree that there is no need to distinguish between actions and hooks, I will combine them into one object. I like the idea of generating webhook ID's on the fly too :+1: 

In regards to the threading, is it a matter of creating a dedicated thread in the existing threadpool and running the webhook server's asyncio event loop within that dedicated thread? I am not that familiar with threading in Python, so I'm kind of learning about it at the same time :)

jneeven commented:
> In regards to the threading, is it a matter of creating a dedicated thread in the existing threadpool and running the webhook server's asyncio event loop within that dedicated thread? I am not that familiar with threading in Python, so I'm kind of learning about it at the same time :)

Exactly! Have a look at what I've done with the scheduler: https://github.com/jneeven/snaketalk/blob/main/snaketalk/bot.py#L53

jneeven commented:
I've just merged #23, which means this is the only item left in the roadmap :tada: Let me know if you want me to do anything, I'd be happy to help :+1:

ttuffin commented:
That's great! You are powering through it :muscle: I may not have much more time to work on it this weekend and I don't want to hold things up, so feel free to ahead with this one as well.

jneeven commented:
> That's great! You are powering through it muscle I may not have much more time to work on it this weekend and I don't want to hold things up, so feel free to ahead with this one as well.

Cool, I'm not sure exactly how much work I'll still do this weekend either, but I'll try to start at the bottom of the list (with the MessageHandler function, the WebhookEvent object etc) so we can hopefully work in parallel a bit. I'll make PRs to this branch (as I've done with https://github.com/ttuffin/snaketalk/pull/1) if I have any updated code :+1:

ttuffin commented:
@jneeven, please let me know if this implementation of running the server on a dedicated thread is correct or not.

jneeven commented:
@ttuffin it looks almost correct, but given the `loop.run_forever()` in the `start` method of the server, I think the `while True` in the thread starter doesn't do anything. If you just remove that part, I assume the code works :+1:

Ideally we'd somehow not use `run_forever`, so that we can periodically check if `self.alive` is still true, but I'm not familiar with  `asyncio` at all so I wouldn't know how to do that without manually scheduling and waiting for a task every iteration of an infinite loop. For now this should work just fine, though we may need to force-close the thread on shutdown.

jneeven commented on snaketalk/driver.py:
```diff
@@ -75,6 +77,13 @@ def run_pending():
+    def start_webhook_server_thread(self):
+        def start_server():
+            logging.info("Webhook server thread started.")
+            while self.alive:
+                self.webhook_server.start()
```
```suggestion
            self.webhook_server.start()
```

ttuffin commented:
@jneeven I looked into this a bit more and I was over complicating things. I should have been using `asyncio.run()` instead of setting up a new event loop manually starting it using `run_forever()`. By using `run()` we can still use the while loop to periodically check if `self.alive` is True and the function also handles the loop closure at the end.

For the second task to add the queue and queue listener to `MessageHandler`, is this to be done using the existing `queue` library that is already imported into `driver.py`? If that's the case then don't we need to run the webhook server thread on a separate target? From my understanding the queues are based on a particular target, and since we want a separate queue for webhook processing does that mean we need run the webhook server on it's own target? I.e. `worker = threading.Thread(target=self.webhook_server.start, name='WebhookServer')`

Again, I am not familiar with either the `threading` and `queue` libraries so apologies for the extra questions.

jneeven commented:
@ttuffin sounds good! Regarding the queue: the `threading.Queue` object is shared between threads, so we can e.g. create a Queue object in the MessageHandler and pass it to the webhook server before it starts running. Anything that is then put into that queue by the webhook server, regardless of the thread it runs on, can be obtained from it by the MessageHandler. 

An open question is how exactly we'll periodically check that queue in the MessageHandler... My initial idea is that we could hopefully share the asyncio event loop that is already used in the websocket listener of the driver. `Driver.connect_websocket` currently runs forever, but we could modify the code to return that event loop so that we can also hook a webhook listener into it, and then start the listeners ourselves.

Again, I'm not very familiar with asyncio, but I think that should be possible. In any case, let me know if you have any better ideas!

jneeven commented:
@ttuffin I think I've now completed the entire list, all in https://github.com/ttuffin/snaketalk/pull/4. The only thing still on the list is the logging, but I'm not sure what exactly you had in mind there? Very basic logging for starting and stopping the server is in place, and each incoming request is automatically logged by `aiohttp`. I think the entire repo could use better and more structured logging, so perhaps we can tackle that in its own PR down the line.

If you're happy with this, I suggest we merge it and start integrating snaketalk into `mmpy_bot` immediately afterwards (provided that CI will pass here, of course).

ttuffin commented:
Great work on this @jneeven! Sorry I haven't been much help; I've been offline most of the weekend. I did have a go at the webhook queuing on Friday night and was looking into `asyncio.queue` since I thought it might be the way to go, but I just got lost down a rabbit hole :)

Agree on the logging, there is probably not much more to add. I added the task so I would come back to it and ensure we had what we needed in that area, and it sounds like we already do.

I fixed one of the linting errors due to unused import but its complaining about isort now. Otherwise I say we ago ahead with the merge!

Also have you tested the webhook server with SSL? If not I can run some checks to see how it goes with a Lets Encrypy cert. This can come later though, and doesn't need to block the merge with mmpy_bot.

jneeven commented:
No problem at all, thanks for getting this started in the first place :sweat_smile:

> Also have you tested the webhook server with SSL? If not I can run some checks to see how it goes with a Lets Encrypy cert. This can come later though, and doesn't need to block the merge with mmpy_bot.

I have not, since the mattermost preview image doesn't seem to support it. Definitely a good thing to check if feasible though.

I'll try to resolve the remaining errors here soon and will then open a PR to mmpy_bot, and then we can make a new to-do list there (and hopefully get some helping hands) :+1:

jneeven commented on snaketalk/function.py:
```diff
@@ -238,7 +238,7 @@ def ensure_response(*args):
         except Exception:
-            logging.error(exc_info=True)
+            logging.exception("Exception occurred: ")
         finally:
```
The `exc_info` was supposed to print the traceback here (but I didn't test it so it was probably broken). This change doesn't print any details about the error at all?
